packages/backend/src/core/discovery/handlers/user/ArrayHandler.ts

```diff
@@ -1,29 +1,100 @@
 import * as z from 'zod'
 
 import { DiscoveryProvider } from '../../provider/DiscoveryProvider'
+import { ContractValue } from '../../types'
 import { Handler, HandlerResult } from '../Handler'
+import { getReferencedName, Reference, resolveReference } from '../reference'
+import { callMethod } from '../utils/callMethod'
+import { toFunctionFragment } from '../utils/toFunctionFragment'
+import { valueToNumber } from '../utils/valueToNumber'
 
 export type ArrayHandlerDefinition = z.infer<typeof ArrayHandlerDefinition>
 export const ArrayHandlerDefinition = z.strictObject({
   type: z.literal('array'),
+  method: z.optional(z.string()),
+  length: z.optional(z.union([z.number().int().nonnegative(), Reference])),
 })
 
 export class ArrayHandler implements Handler {
-  readonly dependencies = []
+  readonly dependencies: string[] = []
+  readonly fragment: FunctionFragment
 
   constructor(
     readonly field: string,
-    private readonly _definition: ArrayHandlerDefinition,
-  ) {}
+    private readonly definition: ArrayHandlerDefinition,
+    abi: string[],
+  ) {
+    const dependency = getReferencedName(definition.length)
+    if (dependency) {
+      this.dependencies.push(dependency)
+    }
+    const method = definition.method ?? field
+    if (method.includes(' ')) {
+      this.fragment = toFunctionFragment(method)
+      if (!isArrayFragment(this.fragment)) {
+        throw new Error('Invalid method abi')
+      }
+    } else {
+      const fragment = abi
+        .filter((x) => x.startsWith(`function ${method}`))
+        .map(toFunctionFragment)
+        .find(isArrayFragment)
+      if (!fragment) {
+        throw new Error(`Cannot find an array method for ${method}`)
+      }
+      this.fragment = fragment
+    }
+  }
+
+  getMethod() {
+    return this.fragment.format(utils.FormatTypes.full)
+  }
 
   async execute(
-    _provider: DiscoveryProvider,
-    _address: EthereumAddress,
+    provider: DiscoveryProvider,
+    address: EthereumAddress,
+    previousResults: Record<string, HandlerResult | undefined>,
   ): Promise<HandlerResult> {
-    return Promise.resolve({
-      field: this.field,
-      value: [],
-    })
+    let length: number | undefined
+    if (this.definition.length !== undefined) {
+      const resolved = resolveReference(this.definition.length, previousResults)
+      length = valueToNumber(resolved)
+    }
+
+    if (length) {
+      const results = await Promise.all(
+        Array.from({ length }).map((_, index) =>
+          callMethod(provider, address, this.fragment, [index]),
+        ),
+      )
+      const errorResult = results.find((x) => x.error)
+      if (errorResult) {
+        return { field: this.field, error: errorResult.error }
+      }
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      const value = results.map((x) => x.value!)
+      return { field: this.field, value }
+    } else {
+      const value: ContractValue[] = []
+      for (let i = 0; ; i++) {
```

Comment:
I'm worried about infinite loop here, especially when running in watch mode. This can happen when someone changes implementation. We would not get any error and it would stop our watch mode.

I suggest we introduce a limit here (200?) and return an error if crossed. Or maybe require a maxLenght to always be passed. Anything that will guarantee we never end up in infinite loop.

Reply:
I've added a `maxLength` field with a `100` default